File: src/test/java/com/raychenon/leetcode/ThreeSumTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

/**
 * User: raychenon
 * Date: 10/7/2020
 * https://leetcode.com/problems/3sum/
 */
public class ThreeSumTest {

    private static ThreeSum threeSum;

    @BeforeClass
    public static void setUp() {
        threeSum = new ThreeSum();
    }

    @Test
    public void example1Test() {
        threeSumEval(Arrays.asList(Arrays.asList(-1, -1, 2), Arrays.asList(-1, 0, 1)), new int[]{-1, 0, 1, 2, -1, -4});
    }

    @Test
    public void example2Test() {
        threeSumEval(Arrays.asList(Arrays.asList(-5, 2, 3)), new int[]{-1, 3, 2, 1, -5, 5});
    }

    @Test
    public void lessThan3elementsTest() {
        threeSumEval(Arrays.asList(), new int[]{-1, 0});
        threeSumEval(Arrays.asList(), null);
    }


    private void threeSumEval(List<List<Integer>> expected, int[] nums) {
        List<List<Integer>> actual = threeSum.threeSum(nums);
        assertThat(expected, is(actual));
    }

}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/ThreeSumTest.kt
package com.raychenon.kotlin.leetcode

import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.Test
import java.util.*

/**
 * User: raychenon
 * Date: 11/7/2020
 * https://leetcode.com/problems/3sum/
 */
class ThreeSumTest {

    @Test
    fun example1Test() {
        threeSumEval(
            Arrays.asList(
                Arrays.asList(-1, -1, 2),
                Arrays.asList(-1, 0, 1)
            ), intArrayOf(-1, 0, 1, 2, -1, -4)
        )
    }

    @Test
    fun example2Test() {
        threeSumEval(
            Arrays.asList(Arrays.asList(-5, 2, 3)),
            intArrayOf(-1, 3, 2, 1, -5, 5)
        )
    }

    @Test
    fun lessThan3elementsTest() {
        threeSumEval(Arrays.asList(), intArrayOf(-1, 0))
    }

    @Test
    fun binarySearchCoverageTest() {
        threeSumEval(
            Arrays.asList(
                Arrays.asList(-2, -1, 3),
                Arrays.asList(-1, -1, 2)
            ), intArrayOf(2, -2, -1, -1, -1, 1, 3)
        )
    }


    private fun threeSumEval(expected: List<List<Int>>, nums: IntArray) {
        val actual = ThreeSum.threeSum(nums)
        MatcherAssert.assertThat(expected, CoreMatchers.`is`(actual))
    }
}